server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ChangelogAction.java

```diff
@@ -133,4 +146,131 @@ private static void writeParameters(JsonWriter json, ChangelogLoader.Change chan
     json.endObject();
   }
 
+
+  /**
+   * @return non-null list of changes, by descending order of date
+   */
+  public List<QProfileChangeDto> load(DbSession dbSession, QProfileChangeQuery query) {
+    List<QProfileChangeDto> dtos = dbClient.qProfileChangeDao().selectByQuery(dbSession, query);
+    List<Change> changes = dtos.stream()
+      .map(Change::from)
+      .collect(MoreCollectors.toList(dtos.size()));
+    completeUserAndRuleNames(dbSession, changes);
+    return dtos;
+  }
+
+  private void completeUserAndRuleNames(DbSession dbSession, List<Change> changes) {
+    Set<String> logins = changes.stream().filter(c -> c.userLogin != null).map(c -> c.userLogin).collect(MoreCollectors.toSet());
+    Map<String, String> userNamesByLogins = dbClient.userDao()
+      .selectByLogins(dbSession, logins)
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(UserDto::getLogin, UserDto::getName));
+
+    Set<RuleKey> ruleKeys = changes.stream().filter(c -> c.ruleKey != null).map(c -> c.ruleKey).collect(MoreCollectors.toSet());
+    Map<RuleKey, String> ruleNamesByKeys = dbClient.ruleDao()
+      .selectDefinitionByKeys(dbSession, Lists.newArrayList(ruleKeys))
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(RuleDefinitionDto::getKey, RuleDefinitionDto::getName));
+
+    changes.forEach(c -> {
+      c.userName = userNamesByLogins.get(c.userLogin);
+      c.ruleName = ruleNamesByKeys.get(c.ruleKey);
+    });
+  }
+
+
+  static class Change {
+    private String key;
+    private String type;
+    private long at;
+    private String severity;
+    private String userLogin;
+    private String userName;
+    private String inheritance;
+    private RuleKey ruleKey;
+    private String ruleName;
+    private final Map<String, String> params = new HashMap<>();
+
+    private Change() {
+    }
+
+    @VisibleForTesting
+    Change(String key, String type, long at, @Nullable String severity, @Nullable String userLogin,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Constructor has 9 parameters, which is greater than 7 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)